Add mattermost notification

ragingpastry opened the pull request:
This PR adds a mattermost notification type. Mattermost is an opensource version of Slack that is bundled with Gitlab. 

The mattermost notification integration uses Incoming Webhooks. These must be enabled in the System Console of Mattermost. If the user wishes to allow Ansible Tower notifications to modify the Icon URL and username of the notification then they must enabled these options as well.

* `url`: The incoming webhook URL that was configured in Mattermost. Notifications will use this URL to POST.
* `username`: Optional. The username to display for the notification.
* `channel`: Optional. Override the channel to display the notification in. Mattermost incoming webhooks are tied to a channel by default, so if left blank then this will use the incoming webhook channel. Note, if the channel does not exist then the notification will error out.
* `icon_url`: Optional. A URL pointing to an icon to use for the notification.

Signed-off-by: Nick Wilburn <senior.crepe@gmail.com>

ragingpastry committed:
Add mattermost notifications

ragingpastry committed:
Remove debugging line

ragingpastry committed:
Add mattermost to notification_system docs

ragingpastry committed:
Add mattermost import to reencrypt

ragingpastry committed:
Remove package-lock.json

matburt commented on awx/ui/templates/ui/index.html:
```diff
@@ -9,14 +9,14 @@
     <script>var $basePath = '{{ STATIC_URL }}'</script>
     
-        <link href="{{ STATIC_URL }}css/vendor.f283b4795bc6921b48e5.css" rel="stylesheet" />
+        <link href="{{ STATIC_URL }}css/vendor.60b270d4310f8d01aab3.css" rel="stylesheet" />
```
Did you intend for these things to change?  I've seen this happen before too @gconsidine any idea what's up here?

ragingpastry replied:
Whoops no I did not. I can revert that.

matburt replied:
I think this is automatically generated by some of our tooling and I definitely want to know why (I just ran into it also).

shanemcd replied:
@matburt I think this started happening after #49 landed.

gconsidine replied:
`index.html` is generated now. The hash in the filename changes when the UI is built for cache busting. I'm adding `index.html` to `.gitignore` in a separate PR.

ragingpastry committed:
Revert unintended changes to ui/index.html

matburt commented:
I don't have a mattermost account so can't verify this at the moment.

I'll take another shot at it in a little while unless someone else can verify.

ragingpastry commented:
I was using this for my testing: `docker run --name mattermost-preview -d --publish 8065:8065 mattermost/mattermost-preview`

mabashian commented on awx/ui/client/src/notifications/shared/type-change.service.js:
```diff
@@ -48,6 +49,12 @@ function (i18n) {
                     break;
+                case 'mattermost':
+                    obj.mattermost_required = true;
+                    obj.mattermost_username_required = false;
```
It looks like username, channel and icon are fields that you want to show when the notification type is `mattermost` but they aren't required.  You shouldn't need these flags in this case.  Since it looks like the mattermost url is required `obj.mattermost_required = true;` is fine.

ragingpastry replied:
Thanks I will make that change. I wasn't too familiar with this so I tried to mirror some of the other notification types

mabashian commented on awx/ui/client/src/notifications/notificationTemplates.form.js:
```diff
@@ -333,6 +333,50 @@ export default ['i18n', function(i18n) {
+            mattermost_username: {
+                label: i18n._('Username'),
+                type: 'text',
+                awRequiredWhen: {
```
If username is not required then you can remove this `awRequiredWhen` block.  The form generator will generate a field that's not required by default.  The same is true for channel and icon url.

mabashian reviewed (changes requested):
Dropped in a few suggestions for simplifying the changes to the UI side of things.  I'm impressed that you picked up the form config syntax so quickly.

matburt commented:
![image](https://user-images.githubusercontent.com/89544/30450041-ac43e710-995e-11e7-91d3-a84bb334b8db.png)

Nice job, this totally works well and has parity with our other notification types.

matburt commented on docs/notification_system.md:
```diff
@@ -99,6 +100,26 @@ Hipchat allows you to create a team with limited users and message history for f
 
+## Mattermost
+
+The mattermost notification integration uses Incoming Webhooks. These must be enabled in the System Console of Mattermost. If the user wishes to allow Ansible Tower notifications to modify the Icon URL and username of the notification then they must enabled these options as well.
```
My one complaint is that we should provide a little bit more information on how to actually turn this on. It wasn't immediately obvious to me where to turn on integrations and then where to find the webhook integrations.  Just a little bit more here would probably do well.

Might also point out that a password isn't required because the notification webhook url itself is the secret.

Nicely done, though... I'm impressed with host fast you turned this around.

ragingpastry replied:
Thanks!

I can definitely beef up the docs a bit, as well as apply the other changes that were pointed out. I'll push up the changes as soon as I have a bit of free time today.

ragingpastry committed:
Remove trailing whitespace

ragingpastry committed:
Remove unnecessary fields from mattermost notification UI files

ragingpastry committed:
Update mattermost notification docs with examples

ragingpastry committed:
Add "Disable SSL Verification" button to mattermost notification

This commit will add a button which will disable SSL notifications
for the mattermost webhook. This is required when using self-signed
certificates.

ragingpastry commented:
Made changes stated above. Also added a checkbox to disable SSL verification for those running mattermost with self-signed certs. I can revert that commit if that should go in another PR / be decided on later

matburt commented:
It's totally fine to have that included in this. If you don't mind, add the git sign off flag to your first, main commit and we'll get this merged.

mabashian reviewed (approved):
lgtm!

matburt commented:
Nicely done @ragingpastry !

wenottingham commented:
@ragingpastry thanks for the contribution!
